File: Baekjoon/18352.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int n, m, k, x;
vector<vector<int>> adj;
vector<int> dist;

void bfs() {
	queue<int> q;
	q.push(x);
	dist[x] = 0;

	while (!q.empty()) {
		int now = q.front();
		q.pop();

		if (dist[now] > k) continue;

		for (int next : adj[now]) {
			if (dist[next] == -1) {
				q.push(next);
				dist[next] = dist[now] + 1;
			}
		}
	}
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	cin >> n >> m >> k >> x;
	adj = vector<vector<int>>(n + 1);
	dist = vector<int>(n + 1, -1);
	for (int i = 0; i < m; i++) {
		int a, b;
		cin >> a >> b;
		adj[a].push_back(b);
	}

	bfs();

	bool find = false;
	for (int i = 1; i <= n; i++)
		if (dist[i] == k) {
			cout << i << '\n';
			find = true;
		}

	if (!find) cout << -1 << '\n';
	return 0;
}

File: Baekjoon/18352.py
import sys
from collections import deque

n, m, k, x = map(int, sys.stdin.readline().rstrip().split())
adj = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, sys.stdin.readline().rstrip().split())
    adj[a].append(b)

distance = [-1] * (n + 1)


def bfs():
    q = deque([x])
    distance[x] = 0

    while q:
        now = q.popleft()
        if distance[now] > k:
            continue

        for next in adj[now]:
            if distance[next] == -1:
                distance[next] = distance[now] + 1
                q.append(next)


bfs()

find = False
for i in range(1, n + 1):
    if distance[i] == k:
        print(i)
        find = True

if not find:
    print(-1)
